feat: Passing `--browser` flag alone will launch after test selection

raygdev opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes [issue 22003](https://github.com/cypress-io/cypress/issues/22003)

### Additional details

- This was a feature request made in the issue listed above
- When the `--browser` flag is passed alone, after selecting the testing type, the selected browser will automatically launch.
- I used the `localSettings` field in the `OpenBrowser` query to check if the browser was set in the cli. That query was passed as a prop to `OpenBrowserList` and used on mount and emits the already defined `launch` function if the browser was set.


### Steps to test

In development `yarn cypress:open --browser <browser-name-or-path>`. Select your testing type in the GUI. After testing type selection, the browser specified will automatically launch. Clicking `switch testing type` will navigate back to select another testing type. When a new type is selected the user will have to select the browser from the UI. If an invalid browser name or path is passed, the cached browser will not launch and the error will show as intended. The user will either have to select from the list of available browsers or relaunch from the cli.

### How has the user experience changed?

Previously, when a user only passed the `--browser` flag, they would still be prompted to select the browser and click the `Start testing in <browser-name>` button to open the browser. The only other option was to pass a testing type flag and browser flag together to automatically launch.

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

raygdev committed:
Add localSettings field to OpenBrowser query

raygdev committed:
Add wasBrowserSetInCLI as OpenBrowserList prop

raygdev committed:
Emit launch on mount if --browser was passed

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cypress-io/cypress?pullRequest=28538) <br/>All committers have signed the CLA.

raygdev committed:
Add entry to cli changelog

raygdev committed:
Correct typo in changelog

raygdev committed:
Add link to issue addressed

raygdev committed:
Add pull request title

raygdev committed:
Check if browser is already open before launch

MikeMcC399 commented on cli/CHANGELOG.md:
```diff
@@ -668,3 +668,9 @@ _Released 1/24/2023_
 - Video output link in `cypress run` mode has been added to it's own line to
   make the video output link more easily clickable in the terminal. Addresses
   [#23913](https://github.com/cypress-io/cypress/issues/23913).
+
+**Features:**
+
+- Passing `--browser` flag alone will launch after test selection
+in the GUI, the browser will automatically launch. Addresses 
+[#22003](https://github.com/cypress-io/cypress/issues/22003)
```
You asked in https://discord.com/channels/755913899261296641/1185979505404289115 why the Semantic Pull Request check is failing

You have added this "Features" section to release `12.4.0` at the bottom of this file.
It would need to be moved to the upcoming new version at the top of this file.

raygdev replied:
🤦‍♂️😂 I didn't even realize there were Semantic version numbers in the file at all! Thank you for pointing that out! When I get back home in a little while, I will make that change. 

Do you have, by chance, any advice on running the tests? And which file I should add my test(s) to?

Thank you for your response!

raygdev committed:
Moved features section to top of file

raygdev committed:
Add reference to PR in changelog

raygdev committed:
Correct unintended completion

raygdev commented on packages/launchpad/src/setup/OpenBrowser.vue:
```diff
@@ -9,7 +9,7 @@
     />
     <OpenBrowserList
       variant=""
-      :was-browser-set-in-c-l-i="query.data.value.localSettings.preferences.wasBrowserSetInCLI"
```
I'm curious if this was a linting rule that was setup. I bypassed this with `git commit -n -m "<message>"` because it was refusing the commit because of an empty commit rule. Now that I'm looking at the way the other props are passed, should it be changed to `was-browser-set-in-cli` instead of `wasBrowserSetInCLI` to follow convention?

marktnoonan replied:
yes this is a linting rule, it would be good to follow the convention of kebab-cased attributes for prop names. Vue magically provides them as camelCase inside the child component, so thing else should have to change.

raygdev commented:
Currently the way the code is, when running the command `yarn cypress:open --browser chrome` it launches as intended. Let's say we provide a browser that isn't on the list like `yarn cypress:open --browser something`. The warning shows as intended, however... the last cached browser still launches even though a warning, about the browser passed not being found, is displayed. I'm wondering what the outcome should be for this. Should we still allow launch of the cached browser for convenience, and let the user dismiss the warning? Or do we want to negate launching altogether if an incorrect or unsupported browser is passed?

raygdev committed:
Merge branch 'cypress-io:develop' into issue-22003

marktnoonan commented:
> Currently the way the code is, when running the command `yarn cypress:open --browser chrome` it launches as intended. Let's say we provide a browser that isn't on the list like `yarn cypress:open --browser something`. The warning shows as intended, however... the last cached browser still launches even though a warning, about the browser passed not being found, is displayed. I'm wondering what the outcome should be for this. Should we still allow launch of the cached browser for convenience, and let the user dismiss the warning? Or do we want to negate launching altogether if an incorrect or unsupported browser is passed?

My suggestion is that if we couldn't find the browser that was passed, we should not open any browser and let you see the warning, then you'd have to either relaunch passing in a different browser name, or pick the browser you want from the list.

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Merge branch 'issue-22003' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Add features to top of changelog

raygdev committed:
Change prop name for convention

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Merge develop and handle conflicts

raygdev committed:
Merge branch 'issue-22003' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Add isValidBrowser

checkes whether a cliBrowser is valid or not and
returns a boolean

raygdev committed:
Add isValidBrowser to schema

raygdev committed:
Use isValid browser

creates a function launchIfBrowserSetInCli that
will launch the browser if a valid browser flag
was passed to the cli

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Add to changelog.md

raygdev commented on packages/graphql/src/schemaTypes/objectTypes/gql-LocalSettings.ts:
```diff
@@ -33,6 +33,23 @@ export const LocalSettingsPreferences = objectType({
+            return false
+          }
+
+          const browser = await ctx._apis.browserApi.ensureAndGetByNameOrPath(ctx.coreData.cliBrowser)
```
I'm not sure if this is acceptable. I am basically using the browserApi to ensure that the cli browser passed is a found browser... It will throw an error otherwise. So I wrap it in a try/catch. If it is caught, the browser is either invalid or something is wrong... so we don't launch if isValidBrowser is false. I believe that `ensureAndGetByNameOrPath` is just [this](https://github.com/cypress-io/cypress/blob/develop/packages/server/lib/browsers/utils.ts#L344) utility function under the hood. I am not sure yet if there are side effects from this but would greatly appreciate some input on if this is ok or not.

raygdev reviewed (commented):
Please let me know if this is acceptable

raygdev commented on packages/launchpad/src/setup/OpenBrowserList.vue:
```diff
@@ -286,9 +286,7 @@ const browserStatus = computed(() => {
-  if (props.wasBrowserSetInCli && !browserStatus.value.open) {
-    emit('launch')
-  }
+  props.launchIfBrowserSetInCli()
```
I passed as a prop and used `onMounted` because when the component mounts, the data inside the function will be available. I am still struggling with making sure it only launches once when mounted. Currently when you navigate back and select the same testing type... it still relaunches based on the conditions in the function. From what I can tell in some of the actions in data context, this should be normal behavior to relauch  IF the testing type is changed otherwise this shouldn't occur. 

I have tried a few different things on the client to get it to work, but they aren't viable solutions and always seem to have an edge case that can't be handle on the client based on the information currently available. Any suggestions for this?

raygdev committed:
Add global launch count to keep track project launch

raygdev committed:
Make global count available to the launchPad

raygdev committed:
Add description for globalLaunchCount

raygdev committed:
Add globalCounnt to schema

raygdev committed:
Remove unused import and remove unused prop

raygdev committed:
Use launch function

raygdev committed:
Import document and use query

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Add to changelog

raygdev committed:
Add to existing features section

raygdev commented on packages/data-context/src/actions/ProjectActions.ts:
```diff
@@ -89,6 +89,13 @@ type SetForceReconfigureProjectByTestingType = {
+   * that launchProject is only called once if
+   * the --browser flag is passed alone.
+   */
+  private globalLaunchCount = 0
```
This is only used to increment in `launchProject` and returned with a getter to use in gql-LocalSettings.ts

raygdev commented on packages/data-context/src/actions/ProjectActions.ts:
```diff
@@ -284,6 +295,7 @@ export class ProjectActions {
     }
 
     await this.api.launchProject(browser, activeSpec ?? emptySpec, options)
+    this.globalLaunchCount++
```
We increment here when the project is launched

raygdev commented on packages/graphql/src/schemaTypes/objectTypes/gql-LocalSettings.ts:
```diff
@@ -50,6 +50,10 @@ export const LocalSettingsPreferences = objectType({
     })
 
+    t.int('globalLaunchCount', {
+      resolve: (_source, _args, ctx) => ctx.actions.project.launchCount,
```
Here I'm just resolving with the `launchCount`

raygdev commented on packages/launchpad/src/setup/OpenBrowserList.vue:
```diff
@@ -285,8 +284,4 @@ const browserStatus = computed(() => {
 const testingTypeIcon = computed(() => props.gql.currentTestingType === 'component' ? TestingTypeComponentIcon : TestingTypeE2EIcon)
 
-onMounted(() => {
-  props.launchIfBrowserSetInCli()
```
I removed this function as a prop and just used in the parent component. When running the tests for `OpenBrowserList` it was failing because this prop was not passed in the tests.

raygdev commented on packages/launchpad/src/setup/OpenBrowser.vue:
```diff
@@ -9,9 +9,7 @@
     />
     <OpenBrowserList
       variant=""
-      :was-browser-set-in-cli="query.data.value.localSettings.preferences.wasBrowserSetInCLI"
```
Removed this prop because it was no longer needed in browser list

raygdev commented on packages/launchpad/src/setup/OpenBrowser.vue:
```diff
@@ -48,7 +42,20 @@ query OpenBrowser {
 }
 `
 
+gql`
```
I created a separate query for LocalSettings

raygdev commented on packages/launchpad/src/setup/OpenBrowser.vue:
```diff
@@ -48,7 +42,20 @@ query OpenBrowser {
+`
+
 const query = useQuery({ query: OpenBrowserDocument })
+const lsQuery = useQuery({ query: OpenBrowser_LocalSettingsDocument, requestPolicy: 'network-only' })
```
Since the values were being cached by the urql client I couldn't get the information to sync. `globalLaunchCount` was always `0` even though the value on the server was updating. 

I notice that this may not be the best choice but couldn't really find a better option. I do feel that since most values are cached anyway, having this as a network request when switching views is negligible. But if anyone has a better idea, I'll be more than happy to make the change.

raygdev commented on packages/launchpad/src/setup/OpenBrowser.vue:
```diff
@@ -113,10 +120,12 @@ const launch = async () => {
 const launchIfBrowserSetInCli = async () => {
-  const isValidBrowser = query.data.value?.localSettings.preferences.isValidBrowser
-  const wasBrowserSetInCli = query.data.value?.localSettings.preferences.wasBrowserSetInCLI
+  await lsQuery
```
I had to await `lsQuery` here to make sure that the data was available without passing as a prop.

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Add to changelog

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Update changelog.md

raygdev committed:
Add setter for launchCount and add tests

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Update changelog

raygdev commented on packages/data-context/src/actions/ProjectActions.ts:
```diff
@@ -138,6 +138,10 @@ export class ProjectActions {
     return this.globalLaunchCount
   }
 
+  set launchCount (count) {
```
I genuinely don't like this, but I don't see any other way to get the tests to pass for the changes I made. I have to provide a setter to reset the launch count before the tests are run. If I don't, the test I made will fail on a second run.

raygdev commented on packages/launchpad/cypress/e2e/choose-a-browser.cy.ts:
```diff
@@ -1,6 +1,12 @@
 describe('Choose a browser page', () => {
+  before(() => {
+    cy.withCtx((ctx, _) => {
+      ctx.actions.project.launchCount = 0
```
I have to set the launch count here before the tests start. I don't feel this is the best way to test it, but I'm out of ideas to get this going.

raygdev commented:
I'm going to open this PR as ready for review so that I can get some feedback on the current code I have. 
- I set up a global launch count in `ProjectActions` to keep track of `launchProject` calls.
- I exposed the `launchCount` with a setter to help with testing.
- I used `gql-LocalSettings` to add a `globalLaunchCount` field and `isValidBrowser` field
- `isValidBrowser` uses what I believe to be a utility function to find a valid browser by name or path and throws otherwise
- I created a new query in `OpenBrowser.vue` that makes a network request (the values were cached which made `globalLaunchCount` always evaluate to `0` even though it was updating on the server) so that the launch function wouldn't be called again when navigating back and selecting a new testing type.
- I used the `OpenBrowser_LocalSettingsDocument` generated from the additional query to keep the request size as small as possible while leaving the other values cached.
- Created a `launchIfBrowserSetInCli` function that uses the three fields needed and uses the already defined `launch` function if the conditions are met and is called on mount of the `OpenBrowser` component.

Since there is already a function that launches with the same values (testingType, wasBrowserSetInCli) I had to create a launch count to keep track of the calls to ensure that both functions wouldn't be called since they rely on the same values. I honestly feel like this is a bit hacky, so I will happily take some advice if changes need to be made. This is my first contribution to a large codebase, so I want to be sure that this is going in the right direction before I code myself into a corner. I will put this back in draft (if need be) after any feedback provided.

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Handle changelog merge conflict

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Merge changes

raygdev committed:
Remove extra bugfix title

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Update changelog

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

jennifer-shehane commented:
@raygdev I'll see if the team has some time to review this next iteration.

raygdev commented:
@jennifer-shehane sounds good! Thank you!

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Update changelog

marktnoonan committed:
Merge branch 'develop' into issue-22003

marktnoonan commented:
@raygdev I hit that "update branch" button for you, it seems like there's now a little bit of cleanup needed in the changelog just FYI

raygdev committed:
Update changelog

raygdev commented:
@marktnoonan Done! Thanks Mark!

ryanthemanuel committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Update changelog

mschile committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge issue-22003

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Update Changelog

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Update changelog

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Update Changelog

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Update changelog

raygdev committed:
Fix changelog for line break error

raygdev committed:
Merge branch 'develop' into issue-22003

alexsch01 commented:
any update on this?

ryanthemanuel commented on packages/launchpad/src/setup/OpenBrowser.vue:
```diff
@@ -106,6 +119,19 @@ const launch = async () => {
+  const wasBrowserSetInCli = lsQuery.data.value?.localSettings.preferences.wasBrowserSetInCLI
+  const globalLaunchCount = lsQuery.data.value?.localSettings.preferences.globalLaunchCount
+
+  if (wasBrowserSetInCli && isValidBrowser && (globalLaunchCount === 0)) {
```
Rather than exposing these new fields and having this business logic in the front end, can we instead expose a dedicated field like `shouldLaunchBrowserFromOpenBrowser` and have this logic abstracted in the server?

raygdev replied:
@ryanthemanuel now that you point that out, that seems more reasonable to do. Thank you! 

As far as creating a new field...
Can/should I still keep it on the `localSettings` object type already defined or should I put it somewhere else?

raygdev committed:
Merge branch 'develop' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Update changelog

raygdev committed:
Refactor to create single field for launching browser

raygdev committed:
Update schema

raygdev committed:
Refactor function to make use of single field

raygdev committed:
Change launch count in beforeEach hook instead

raygdev committed:
Merge branch 'issue-22003' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Clean up await in function

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Resolve Changelog conflict

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

marktnoonan committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Update changelog

raygdev committed:
Merge branch 'develop' into issue-22003

marktnoonan committed:
Merge branch 'develop' into issue-22003

marktnoonan commented on packages/launchpad/src/setup/OpenBrowser.vue:
```diff
@@ -106,6 +117,16 @@ const launch = async () => {
 }
 
+const launchIfBrowserSetInCli = async () => {
+  const shouldLaunchBrowser = (await lsQuery).data.value?.localSettings.preferences.shouldLaunchBrowserFromOpenBrowser
```
I suppose it's possible that the query could somehow return without everything that's expected being populated, so we could do more optional chaining here to be more resilient to that.

```suggestion
  const shouldLaunchBrowser = (await lsQuery).data.value?.localSettings?.preferences?.shouldLaunchBrowserFromOpenBrowser
```

raygdev replied:
That makes sense! Thanks @marktnoonan! Added the suggested change.

marktnoonan reviewed (changes requested):
Tested this locally and everything works as expected, just a one-line change request.

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Add additional optional chaining for resiliency

marktnoonan reviewed (approved):
This looks good to me, tested locally and everything checks out. To your question about `localSettings` for `shouldLaunchBrowserFromOpenBrowser` I think that's a fine place because it's close to `wasBrowserSetInCLI` and makes a much sense as any other place.

raygdev commented:
I do have a question around testing. 

In [choose-a-browser.cy.ts](https://github.com/cypress-io/cypress/blob/develop/packages/launchpad/cypress/e2e/choose-a-browser.cy.ts#L39)... on this line, I wanted to assert that the `launchProject` function has not been called to reassure (in testing) that it doesn't launch the browser when an error occurs for a browser that doesn't exist or was not found. I can manually test and know that it doesn't, but its not the same as having a test case around the behavior.

When I run the test for what I have, it shows the function has been called. Maybe it's because of something around the mocks or stubs that I don't quite understand. Would anyone be able to point me in the right direction to get this test  in a working state for asserting that? Or point me in a direction to find another solution for a standalone test?

raygdev commented:
> This looks good to me, tested locally and everything checks out. To your question about `localSettings` for `shouldLaunchBrowserFromOpenBrowser` I think that's a fine place because it's close to `wasBrowserSetInCLI` and makes a much sense as any other place.

Thank you for clarifying @marktnoonan! I appreciate it!

raygdev committed:
Merge branch 'develop' into issue-22003

marktnoonan commented:
@raygdev I think you hit on an actual, but silent, bug related to the code that runs when a user dismisses the welcome page. In [this function in Main.vue](https://github.com/cypress-io/cypress/blob/687eb6029f4a0cb22d411b12b3fd16df973c4da0/packages/launchpad/src/Main.vue#L249), we attempt to launch the project after the welcome page is dismissed, if testing type and browser are set. If the provided browser doesn't exist, we still end up in the correct state.

You can get your test to pass by changing `wasBrowserSetInCLI` to your more precise `shouldLaunchBrowserFromOpenBrowser` inside that function, so I suggest making that change in your PR and adding the test you mentioned.

After that change, you should be able to stub `launchProject` like elsewhere in that choose-a-browser spec file:

```js
cy.withCtx((ctx, o) => {
  // stub launching project since we have `--browser --testingType` here
  o.sinon.stub(ctx._apis.projectApi, 'launchProject').resolves()
})
```

And assert that it is never called by the end of the test:

```js
cy.withRetryableCtx((ctx, o) => {
  expect(ctx._apis.projectApi.launchProject).to.not.have.been.called
})
```

This may have been exactly what you were trying already, just providing it as a reference anyway.

raygdev committed:
Use more precise browser launching query to fix silent bug

raygdev committed:
Assert that launchProject hasn't been called when browser not found

raygdev commented:
@marktnoonan that was exactly how I was trying to assert... but it kept failing. I wasn't sure if it was an actual bug or my misunderstanding! I would have battled for a little while to find that. Thank you for looking into it. Those changes have been made.

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Update changelog

raygdev committed:
Merge branch 'issue-22003' of https://github.com/raygdev/cypress into issue-22003

raygdev committed:
Update changelog

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Update changelog

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

raygdev committed:
Merge branch 'develop' into issue-22003

jennifer-shehane commented on cli/CHANGELOG.md:
```diff
@@ -16,6 +16,8 @@ _Released 8/27/2024 (PENDING)_
 - Added a `CYPRESS_SKIP_VERIFY` flag to enable suppressing Cypress verification checks. Addresses [#22243](https://github.com/cypress-io/cypress/issues/22243).
 - Updated the protocol to allow making Cloud API requests. Addressed in [#30066](https://github.com/cypress-io/cypress/pull/30066).
 
+- Passing the browser without the testing type (i.e. `cypress open --browser <browser-name-or-path>`) will now directly launch the browser after the testing type is selected. Addresses [#22003](https://github.com/cypress-io/cypress/issues/22003). Addressed in [#28538](https://github.com/cypress-io/cypress/pull/28538).
```
```suggestion
- Updated the protocol to allow making Cloud API requests. Addressed in [#30066](https://github.com/cypress-io/cypress/pull/30066).
- Passing the browser without the testing type (i.e. `cypress open --browser <browser-name-or-path>`) will now directly launch the browser after the testing type is selected. Addresses [#22003](https://github.com/cypress-io/cypress/issues/22003). Addressed in [#28538](https://github.com/cypress-io/cypress/pull/28538).
```

jennifer-shehane committed:
Update cli/CHANGELOG.md

jennifer-shehane committed:
Merge branch 'develop' into issue-22003

jennifer-shehane commented:
@raygdev Trying to get this in for today's release. If the tests seem to be passing - we'll get it in. PLEASE do not merge in develop. We'll handle keeping the branch up to date.

raygdev commented:
@jennifer-shehane sounds good! Thank you for your help! Sorry it took me a while to flesh all of it out.

Should I update the cypress documentation for this as well? I haven't done it yet.

jennifer-shehane commented:
@raygdev You could open a PR against this branch: https://github.com/cypress-io/cypress-documentation/pull/5898 We could also update it later, so it's not a big deal if it doesn't make it into that branch in time.
